From bc5c0be1f5d3ad4c074b3951c171d7464de69781 Mon Sep 17 00:00:00 2001
From: Vassil Vassilev <v.g.vassilev@gmail.com>
Date: Mon, 2 Nov 2020 11:48:28 +0000
Subject: [PATCH 098/110] Reset the LexLevel when parsing recursively.

This can happen when using the LookupHelper.

See llvm-mirror/clang@444665e.
---
 include/clang/Lex/Preprocessor.h | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/include/clang/Lex/Preprocessor.h b/include/clang/Lex/Preprocessor.h
index 6b393e9bea..7258437edd 100644
--- a/include/clang/Lex/Preprocessor.h
+++ b/include/clang/Lex/Preprocessor.h
@@ -1340,6 +1340,7 @@ public:
     TokenLexer* SavedCurTokenLexer;
     const DirectoryLookup *SavedCurDirLookup;
     enum CurLexerKind SavedCurLexerKind;
+    unsigned SavedLexLevel;
 
   public:
     CleanupAndRestoreCacheRAII(Preprocessor &PP)
@@ -1350,7 +1351,8 @@ public:
         SavedCurPPLexer(PP.CurPPLexer),
         SavedCurTokenLexer(PP.CurTokenLexer.release()),
         SavedCurDirLookup(PP.CurDirLookup),
-        SavedCurLexerKind(PP.CurLexerKind)
+        SavedCurLexerKind(PP.CurLexerKind),
+        SavedLexLevel(PP.LexLevel)
     {
       PP.CachedLexPos = 0;
       PP.CachedTokens.clear();
@@ -1360,6 +1362,7 @@ public:
       PP.CurTokenLexer.reset(0);
       PP.CurDirLookup = 0;
       PP.CurLexerKind = CLK_CachingLexer;
+      PP.LexLevel = 0;
     }
 
     void pop() {
@@ -1375,6 +1378,7 @@ public:
       PP.CurTokenLexer.reset(SavedCurTokenLexer);
       PP.CurDirLookup = SavedCurDirLookup;
       PP.CurLexerKind = SavedCurLexerKind;
+      PP.LexLevel = SavedLexLevel;
 
       SavedCachedLexPos = 0;
       SavedCachedTokens.clear();
@@ -1384,6 +1388,7 @@ public:
       SavedCurTokenLexer = 0;
       SavedCurDirLookup = 0;
       SavedCurLexerKind = (enum CurLexerKind)~0U;
+      SavedLexLevel = ~0U;
     }
 
     ~CleanupAndRestoreCacheRAII() {
-- 
2.29.2

